### PR TITLE
SNOW-1063723: [Local Testing] Add support for concat and concat_ws

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1063723

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR adds support for the concat and concat_ws functions. I also imported pandas from the connector instead of in each individual mock function. This pandas import only works if pandas is installed and will throw an error if used without pandas installed.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_functions.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 import pytz
 
 import snowflake.snowpark
+from snowflake.connector.options import pandas
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:20]: Thanks for fixing this, ideally we can enforce this with some linting in the precommit.

File: src/snowflake/snowpark/mock/_functions.py [hunk 2]
```diff
@@ -1561,3 +1552,31 @@ def mock_current_database():
...
+
+@patch("concat")
+def mock_concat(*columns: ColumnEmulator) -> ColumnEmulator:
+    if len(columns) < 1:
+        raise ValueError("concat expects one or more column(s) to be passed in.")
+    pdf = pandas.concat(columns, axis=1).reset_index(drop=True)
+    result = pdf.T.apply(
+        lambda c: None if c.isnull().values.any() else c.astype(str).str.cat()
+    )
+    result.sf_type = ColumnType(StringType(), False)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1565]: nullable should be `result.hasnans`?

File: src/snowflake/snowpark/mock/_functions.py [hunk 3]
```diff
@@ -1561,3 +1552,31 @@ def mock_current_database():
...
+        raise ValueError(
+            "concat_ws expects a seperator column and one or more value column(s) to be passed in."
+        )
+    pdf = pandas.concat(columns, axis=1).reset_index(drop=True)
+    result = pdf.T.apply(
+        lambda c: None
+        if c.isnull().values.any()
+        else c[1:].astype(str).str.cat(sep=c[0])
+    )
+    result.sf_type = ColumnType(StringType(), False)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1581]: Ditto.

**Review by sfc-gh-stan — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/mock/_functions.py (hunk 2 above)
Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_functions.py:1565]: Good catch. Originally I thought the result dropped those rows, but they do remain.

File: CHANGELOG.md [hunk 4]
```diff
@@ -25,6 +25,8 @@
   - snowflake.snowpark.functions
     - current_database
     - current_session
+    - concat
+    - concat_ws
```
Comment (sfc-gh-aling) [inline CHANGELOG.md:29]: nit: you probably need to rebase the main branch and move the changelog to 1.16?

**Review by sfc-gh-aling — APPROVED** [review]

**Review by sfc-gh-stan — APPROVED** [review]

**Pull request merged**